_posts/2023-08-29-what-in-the-ml-is-going-on-around-here.md

```diff
@@ -0,0 +1,473 @@
+GET /_plugins/_ml/tasks/GeLGTIoBKue4OlrZmck7
+```
+
+The response was not quite what I expected.
+
+```json
+{
+  "task_type": "REGISTER_MODEL",
+  "function_name": "TEXT_EMBEDDING",
+  "state": "FAILED",
+  "worker_node": [
+    "_QJb--HRS2-7lfq5DCWMiQ"
+  ],
+  "create_time": 1693505198395,
+  "last_update_time": 1693505199947,
+  "error": "model content changed",
+  "is_async": true
+}
+```
+
+Crap on a crap cracker. 
+
+### Lesson Learned 1: Model Content Hash Value Field
+
+It clearly failed, but `model content changed` isn't very helpful of an error message. I was missing something. 
+
+...*fast forward montage through hours of scanning through `MLModel.java` and pleading for help on our [public slack channel](...)*...
+
+It turns out the example was bad. I was missing a field called `model_content_hash_value`. It's the `sha256` checksum of the zipfile. If you ever want to use a model that's not in the documentation examples, you'll need to calculate the `sha256` checksum of the zipfile and provide that as the `model_content_hash_value` field. Unfortunately there wasn't any documentation for this field. I filed **[Issue 4966](https://github.com/opensearch-project/documentation-website/issues/4966)**.
+
+After downloading the model zip file, I was able to calculate it on my own like so: 
+
+```shell
+$ shasum -a 256 all-MiniLM-L6-v2_torchscript_sentence-transformer.zip
+9376c2ebd7c83f99ec2526323786c348d2382e6d86576f750c89ea544d6bbb14  all-MiniLM-L6-v2_torchscript_sentence-transformer.zip
+```
+
+The lesson? Your API calls to register models via URL require some assembly. This field does not seem to be required when registering a ["pre-trained" model.  ](https://opensearch.org/docs/latest/ml-commons-plugin/pretrained-models/#supported-pretrained-models).
+
+### Lesson Learned 2: Model Groups
+
+I had attempted to upload a model without passing in a model group id. OpenSearch does something for you when this happens. A model group is created for you, with the same name that you provided in your call. In my case, it was `all-MiniLM-L6-v2`. So, the next time I tried to upload that model, it kept telling me the name was taken by a particular model id, so I used the API to search for all of the models available. It wasn't there. What *was* there was a model group that I was able to delete. I used the API to delete all the tasks, models, and model groups I just made so I could start with a fresh slate. Make sure you follow the Tasks, Models and Model Groups side quest to make sure you can organize to your own level of comfort.
+
+The error message was confusing, so I filed [issue 1289](https://github.com/opensearch-project/ml-commons/issues/1289).
+
+
+### Side Quest: Tasks, Models, and Model Groups
+
+Many of the API operations we'll end up using return a task id, model id, or model group id that you have to then look up. Since we're only using the tools that come out of the box, here are the dev console examples that are likely of most help.
```

Comment:
Ensure that you're consistent in your voice throughout the post. Here we switch to "we'll" and "we're", when you've previously been speaking as "I". In general, for procedures or instructions, we frame action as being taken by the user (“Then you can stop the container...”) rather than the writer (“We also have to stop the container...”). Because you're writing this post from a first-person perspective, you'll probably want to stick to mostly "I" and "you" while avoiding "we" and "our" (these appear more frequently later in the post).

Reply:
Got it. This is likely something with which I'll need your help frequently until I can incorporate the habit.